stats.c

```diff
@@ -2426,80 +2433,45 @@ int main_stats(int argc, char *argv[])
-                if (reglist) {
-
-                    hts_itr_multi_t *iter = sam_itr_regions(bam_idx, info->sam_header, reglist, regcount);
-                    if (iter) {
-
-                        if (!targets) {
-                            all_stats->nchunks = argc-optind;
-                            if ( replicate_regions(all_stats, iter) )
-                                fprintf(stderr, "Replications of the regions failed.");
-                        }
+        if (bam_idx) {
+            hts_itr_multi_t *iter = sam_itr_regarray(bam_idx, info->sam_header, &argv[optind], argc - optind);
+            if (iter) {
+                if (!targets) {
+                    all_stats->nchunks = argc-optind;
+                    if (replicate_regions(all_stats, iter))
+                        fprintf(stderr, "Replications of the regions failed\n");
+                }
 
-                        if ( all_stats->nregions && all_stats->regions ) {
-                            while (sam_itr_multi_next(info->sam, iter, bam_line) >= 0) {
-                               if (info->split_tag) {
-                                   curr_stats = get_curr_split_stats(bam_line, split_hash, info, targets);
-                                   collect_stats(bam_line, curr_stats, read_pairs);
-                               }
-                               collect_stats(bam_line, all_stats, read_pairs);
-                            }
+                if ( all_stats->nregions && all_stats->regions ) {
+                    while (sam_itr_next(info->sam, iter, bam_line) >= 0) {
+                        if (info->split_tag) {
+                            curr_stats = get_curr_split_stats(bam_line, split_hash, info, targets);
+                            collect_stats(bam_line, curr_stats, read_pairs);
                         }
-
-                        hts_itr_multi_destroy(iter);
-                    } else {
-                       fprintf(stderr, "Creation of the region iterator failed.");
-                       hts_reglist_free(reglist, regcount);
+                        collect_stats(bam_line, all_stats, read_pairs);
                     }
-                } else {
-                    fprintf(stderr, "Creation of the region list failed.");
                 }
-
-                hts_idx_destroy(bam_idx);
+                hts_itr_multi_destroy(iter);
             } else {
-                fprintf(stderr, "Random alignment retrieval only works for indexed BAM files.\n");
+                fprintf(stderr, "Multi-region iterator could not be created\n");
             }
-
-            bed_destroy(region_hash);
+            hts_idx_destroy(bam_idx);
         } else {
-            fprintf(stderr, "Creation of the region hash table failed.\n");
+            if (has_index_file)
+                fprintf(stderr, "Invalid index file '%s'\n", bam_idx_fname);
+            fprintf(stderr, "Random alignment retrieval only works for indexed files\n");
```

Comment:
Missing else?
Also need to return an error.

Reply:
> Missing else?

No, just that, when the user specifies the index file, a hint could be given about the validity of the file.